### ISSUE TITLE
A saves_per_epoch that creates a fraction may create training that don´t have checkpoints

### ISSUE THREAD

**Issue by bratao** [axolotl-ai-cloud/axolotl#1642]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

If I set the saves_per_epoch to 6, to save 6 checkpoints for each epoch.

### Current behaviour

I set the saves_per_epoch to 6. After a (costly) run it was not saving any checkpoints.
Debugging I realized that axolotl used the formula:
`save_steps = 1.0 / (cfg.saves_per_epoch * cfg.num_epochs)`

However 1/6 = 0.1666666666666667, and during the training this exact value was not archived in the code that check if it is time to save the checkpoint:

```
        if (
            args.save_strategy == IntervalStrategy.STEPS
            and args.save_steps > 0
            and state.global_step % args.save_steps == 0
        ):
            control.should_save = True
```

### Steps to reproduce

Set saves_per_epoch to 6

### Possible solution

The callback check should compare using a approximate equality, like [math.isclose()](https://docs.python.org/3/library/math.html#math.isclose)
We should also add an extra variable to store the lasted saved epoch to avoid saving consecutive checkpoints because of the approximation

### Which Operating Systems are you using?

- [ ] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

main

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by winglian** [comment]
hey @bratao, is the issue that it saved no checkpoints at all or that it keeps only the last 4 checkpoints?

**Comment by bratao** [comment]
@winglian it saved no checkpoint at all

**Comment by winglian** [comment]
@bratao see #1643 for the fix and explanation.

**Comment by bratao** [comment]
Thank you!
